pkg/session/session.go

```diff
@@ -919,6 +919,7 @@ func (s *session) CommitTxn(ctx context.Context) error {
 	r, ctx := tracing.StartRegionEx(ctx, "session.CommitTxn")
 	defer r.End()
 
+	s.setLastTxnInfoBeforeTxnEnd()
```

Comment:
How about putting it in `onTxnEnd`?

Reply:
Seems onTxnEnd，the txn info has been cleared

Reply:
After this line, in `doCommitWithRetry` -> `doCommit`, it seems it still have possibility to finally enter a rolled-back state or encounter error. Perhaps it's better to consider put it in a later place where the commit/rollback result is finally determined.

Reply:
Seems if a txn comitted or fails to commit, the CommitHook will update it: https://github.com/pingcap/tidb/blob/c71ebaf73ed9620745f327ad147f27fc06d73297/pkg/sessiontxn/isolation/base.go#L506

`setLastTxnInfoBeforeTxnEnd` in `CommitTxn` will be overridden at this time. So `setLastTxnInfoBeforeTxnEnd` is only affects some readonly txns.